### PR TITLE
Replace `NullRelation` with a `#null?` state for `Relation`

### PR DESCRIPTION
**Pull request by Empact**
In all cases where `ActiveRecord` currently generates a query with
 `WHERE 1=0`, this will instead avoid the query and return a `#null?` `Relation`,
ala `.none`. This covers cases such as `.where('1=0')`, `.where(id: [])`,
and `.limit(0)`.

This line of thought was inspired by the need for #19349, which is
caused `NullRelation` state leaking across `#reset` boundaries. This
alternative is tidier in that it is 100% resetable, and does not
require hacky cache clearing.

Note this patch still needs documentation, and perhaps other refinement,
but the functionality works as is. This patch would be improved if
`Arel::Nodes::In` responded to a method to indicate when it was a none
predicate - see `WhereClause#none_predicate?`, but I figured I'd get you
guys' opinion on it first.

### CODE REVIEW

**Comment by Empact** [comment]
Note the null state is held in both limit_value and where_value. If we want to protect people it will be easy to add another value to the relation that can not be cleared by e.g. excepting the limit and rewhereing.

File: activerecord/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,13 @@
+*   Replace `NullRelation` with a `#null?` state for `Relation`, triggered by
+    guarranteed empty results, including `.where('1=0')`, `.where(id: [])`,
```
Comment (Amit-Thawait) [inline activerecord/CHANGELOG.md]: typo here : `guarranteed` should be `guaranteed`

**Comment by rafaelfranca** [comment]
I don't know. I don't see why adding a lot of `if` in the code is better that using polymorphism for this case. If the problem is that we can't clean the `extending` call when calling `reset` why not just fix `reset`?

**Comment by Empact** [comment]
@rafaelfranca well, there are a number of cases in which a relation is guaranteed to return 0 results, and currently we're only treating the `.none` relations as Null, whereas this covers a number of other cases  consistently.

As far as I can tell, the problem with `extending` is fundamental, as there's no way to back a module out of the call chain. That could be an interesting patch to the language but I don't think it's possible with current ruby.

I took some care to make the `#null?` call fast - I'll look into benchmarking the overall effect.

**Comment by matthewd** [comment]
Yeah, I don't really see the point of this.

`none` has special behaviour so that people can use it instead of proliferating nil checks, or something, without concern that they're now introducing an always-useless query. That is: you use it in places where your default alternative was never to run a query, but rather to use some other means of signalling "no result" up the stack.

For anything else, where you'll *generally* be running the query, but sometimes, if we pre-evaluate enough of the conditions, we can work out it won't have any results... if it's so clearly empty, the DBMS will already know how to short-circuit any real work. We don't seem to be gaining benefit, to justify even a small amount of effort/code.

**Comment by Empact** [comment]
I can tell you that `WHERE 1=0` checks are fairly common in our codebase, perhaps due to our direct use of `ActiveRecord::Associations::Preloader`.

In this PR we're not doing any difficult work to discover that the relation doesn't have a result, just checking the existing `_values` on the relation, which are straightforward to detect.

Yes, this is an optimization of a corner case, but it's one that we're seeing and the resulting refactor is IMO semantically clean - arguably cleaner than the status quo, in which we have several ways of handling these cases and detecting them.

IMO it's cleaner to have the relation know that it's not going to produce a value, rather than generate an explicit command to the database that says "don't generate any values" then send that command to the DB and wait for it to come back.

**Comment by Empact** [comment]
An excerpt from our slack channel:

screen shot 2015-03-17 at 11 31 08 am

**Comment by sgrif** [comment]
I'm inclined to agree with the others here, I just don't see the value here. It seems fairly simple to eliminate those `WHERE 1=0` queries at the application level. Although I wouldn't mind seeing improvement in our elimination of them in Active Record, I really don't think this approach is the right way to go about it.

**Comment by Empact** [comment]
@sgrif Please elaborate on how you would like to see them go. e.g. do you think #19371 is more appropriate? What do you make of the extend leakage shown by #19349?

**Comment by sgrif** [comment]
I'm not entirely convinced it's a big enough problem that Rails needs to solve it if it takes a significant amount of code. There's plenty of edge cases #19371 doesn't cover, and it's already affecting more places than I'd like. I'm looking more closely at the extend leakage PR -- But really I'm not sure why `NullRelation` is a module that we extend in, in the first place, as opposed to an entirely different class.

**Pull request closed without merging**